### PR TITLE
Upgrade to Smithy IDL 2.0

### PR DESCRIPTION
**Pull request by syall**
For changes to files under the `/model/` folder, and manual edits to autogenerated code (e.g. `/service/s3/api.go`) please create an Issue instead of a PR for those type of changes.

N/A

If there is an existing bug or feature this PR is answers please reference it here.

N/A

---

# **Dependent on PR aws/smithy-go#376: See the PR for much more details about changes, testing, and open issues**

*Description of changes:*

Upgrade to Smithy IDL 2.0:

- Upgrade `smithyVersion` to 1.23.1
- Extract Smithy Gradle Plugin to `smithyGradleVersion` to 0.6.0
- Add support for `jsonName` serialization/deserialization for union shapes

Update generated protocol tests.

---

*Testing:*

File Structure:

```tree
.
├── aws-sdk-go-v2
└── smithy-go
```

1. smithy-go: Ran `make smithy-clean smithy-build smithy-publish-local smithy-clean`.
2. aws-sdk-go-v2: Ran `make`
3. aws-sdk-go-v2: Verified expected diffs in protocol tests and services

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by syall** [comment]
`Codegen Tests / SDK Codegen Test` CI Action failed:

```text
Projection $EXAMPLE failed: java.lang.NoSuchMethodError: 'software.amazon.smithy.model.shapes.NamedMembersShape$Builder 
```

This is a known issue due to the Smithy version mismatch between `aws-sdk-go-v2` (this PR using 1.23.0) and `smithy-go` (the CI action using 1.21.0).

Testing details can be found in aws/smithy-go#376.

**Comment by jasdel** [comment]
Looks like there are a handful of protocol tests that are failing with Smithy 1.23.1 (not sure if this was in 1.23.0 as well)

```
--- FAIL: TestClient_PostUnionWithJsonName_awsRestjson1Serialize (0.00s)
    --- FAIL: TestClient_PostUnionWithJsonName_awsRestjson1Serialize/PostUnionWithJsonNameRequest1 (0.00s)
        api_op_PostUnionWithJsonName_test.go:174: expect body equal, got JSON documents not equal, JSON mismatch (-expect +actual):
              map[string]any{
                "value": map[string]any{
            -           "FOO": string("hi"),
            +           "foo": string("hi"),
                },
              }
    --- FAIL: TestClient_PostUnionWithJsonName_awsRestjson1Serialize/PostUnionWithJsonNameRequest2 (0.00s)
        api_op_PostUnionWithJsonName_test.go:174: expect body equal, got JSON documents not equal, JSON mismatch (-expect +actual):
              map[string]any{
                "value": map[string]any{
            -           "_baz": string("hi"),
            +           "baz":  string("hi"),
                },
              }
--- FAIL: TestClient_PostUnionWithJsonName_awsRestjson1Deserialize (0.00s)
    --- FAIL: TestClient_PostUnionWithJsonName_awsRestjson1Deserialize/PostUnionWithJsonNameResponse1 (0.00s)
        api_op_PostUnionWithJsonName_test.go:302: expect c.ExpectResult value match:
            values do not match
              &awsrestjson.PostUnionWithJsonNameOutput{
            -   Value: &types.UnionWithJsonNameMemberFoo{Value: "hi"},
            +   Value: &types.UnknownUnionMember{Tag: "FOO"},
                ... // 2 ignored fields
              }
    --- FAIL: TestClient_PostUnionWithJsonName_awsRestjson1Deserialize/PostUnionWithJsonNameResponse2 (0.00s)
        api_op_PostUnionWithJsonName_test.go:302: expect c.ExpectResult value match:
            values do not match
              &awsrestjson.PostUnionWithJsonNameOutput{
            -   Value: &types.UnionWithJsonNameMemberBaz{Value: "hi"},
            +   Value: &types.UnknownUnionMember{Tag: "_baz"},
                ... // 2 ignored fields
              }
              ```

**Comment by syall** [comment]
> Looks like there are a handful of protocol tests that are failing with Smithy 1.23.1 (not sure if this was in 1.23.0 as well). The failing test seem to all center around jsonNaming of members.

Will update the PRs (#1806, aws/smithy-go#376) with fixes to upgrade to Smithy 1.23.1 instead of Smithy 1.23.0.

**Comment by syall** [comment]
> Looks like there are a handful of protocol tests that are failing with Smithy 1.23.1 (not sure if this was in 1.23.0 as well). The failing test seem to all center around jsonNaming of members.

The protocol tests should be fixed with the latest changes in aws-sdk-go-v2:

```diff
diff --git a/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeDeserVisitor.java b/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeDeserVisitor.java
index 96d06a41b1..2bba51d5dd 100644
--- a/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeDeserVisitor.java
+++ b/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeDeserVisitor.java
@@ -231,8 +231,9 @@ public class JsonShapeDeserVisitor extends DocumentShapeDeserVisitor {
                             symbolProvider.toMemberName(member),
                             symbol.getNamespace()
                     ).build();
+                    String serializedMemberName = getSerializedMemberName(member);
 
-                    writer.openBlock("case $S:", "", member.getMemberName(), () -> {
+                    writer.openBlock("case $S:", "", serializedMemberName, () -> {
                         writer.write("var mv $P", targetSymbol);
                         target.accept(getMemberDeserVisitor(member, "mv"));
 
diff --git a/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeSerVisitor.java b/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeSerVisitor.java
index 2485913ddd..311318cd3f 100644
--- a/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeSerVisitor.java
+++ b/codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeSerVisitor.java
@@ -235,9 +235,10 @@ final class JsonShapeSerVisitor extends DocumentShapeSerVisitor {
                         symbolProvider.toMemberName(member),
                         symbol.getNamespace()
                 ).build();
+                String serializedMemberName = getSerializedMemberName(member);
 
                 writer.openBlock("case *$T:", "", memberSymbol, () -> {
-                    writer.write("av := object.Key($S)", member.getMemberName());
+                    writer.write("av := object.Key($S)", serializedMemberName);
                     target.accept(getMemberSerVisitor(member, "uv.Value", "av"));
                 });
             }
```

File: service/evidently/api_op_UpdateExperiment.go [hunk 1]
```diff
@@ -64,6 +64,9 @@ type UpdateExperimentInput struct {
 
 	// Removes a segment from being used in an experiment. You can't use this parameter
 	// if the experiment is currently running.
+	//
+	// Deprecated: Use Boolean instead, and add the @default trait to structure members
+	// that targets this shape
 	RemoveSegment bool
```
Comment (jasdel) [inline service/evidently/api_op_UpdateExperiment.go:70]: This blurb about `Deprecated` is really odd and doesn't make sense from a user of the SDK's point of view. From the wording this looks like some model or Smithy 2.0 related issue that shouldn't be present on public documentation of a member.

Comment (syall) [reply service/evidently/api_op_UpdateExperiment.go:70]: I originally thought this diff was acceptable, but looking at the other smithy-* repos, it looks like that these comments are suppressed for Prelude shapes, e.g. in awslabs/smithy-typescript#581.

Comment (syall) [reply service/evidently/api_op_UpdateExperiment.go]: Fixed in https://github.com/aws/smithy-go/pull/376/commits/5a226b6c3d049a0f8b7c9c5eae513981d8b0d680#diff-8a690a7a6321f66f8c1bf14b573baecaa233b532129c36d51ae6e301e36691e9

**Review by jasdel — APPROVED** [review]
Looks good, thanks for making these updates, and correcting the generated docs.

**Comment by jasdel** [comment]
The golangci issue is a false negative, and the codegen test were verified locally. They cannot pass until the smithy-go change is merged in.

**Pull request merged**